libsql/src/hrana/mod.rs

```diff
@@ -198,33 +224,38 @@ impl RowsInner for Rows {
         }))
     }
 
-    fn column_count(&self) -> i32 {
-        self.cols.len() as i32
+    pub fn column_count(&self) -> i32 {
+        self.cursor_step.cols().len() as i32
     }
 
-    fn column_name(&self, idx: i32) -> Option<&str> {
-        self.cols
+    pub fn column_name(&self, idx: i32) -> Option<&str> {
+        self.cursor_step
+            .cols()
             .get(idx as usize)
             .and_then(|c| c.name.as_ref())
             .map(|s| s.as_str())
     }
+}
 
-    fn column_type(&self, idx: i32) -> crate::Result<ValueType> {
-        let row = match self.rows.get(0) {
-            None => return Err(crate::Error::QueryReturnedNoRows),
-            Some(row) => row,
-        };
-        let cell = match row.get(idx as usize) {
-            None => return Err(crate::Error::ColumnNotFound(idx)),
-            Some(cell) => cell,
-        };
-        Ok(match cell {
-            proto::Value::Null => ValueType::Null,
-            proto::Value::Integer { .. } => ValueType::Integer,
-            proto::Value::Float { .. } => ValueType::Real,
-            proto::Value::Text { .. } => ValueType::Text,
-            proto::Value::Blob { .. } => ValueType::Blob,
-        })
+#[async_trait::async_trait]
+impl<S> RowsInner for HranaRows<S>
+where
+    S: Stream<Item = std::io::Result<Bytes>> + Send + Sync + Unpin,
+{
+    async fn next(&mut self) -> crate::Result<Option<super::Row>> {
+        self.next().await
+    }
+
+    fn column_count(&self) -> i32 {
+        self.column_count()
+    }
+
+    fn column_name(&self, idx: i32) -> Option<&str> {
+        self.column_name(idx)
+    }
+
+    fn column_type(&self, _idx: i32) -> crate::Result<ValueType> {
+        todo!()
```

Comment:
are we tracking this todo in an issue somewhere?

Reply:
I don't think so, I've only started a discussion on Slack about this problem.

Reply:
I changed that to work on the best effort basis - when we iterate over the first row, we'll infer and cache the column types from it. If user will try to call the `column_type` prior fetching first row, we'll return an error. Atm. it's the best that we can do.